File: IPWorks MQ Samples/Amazon SNS Notification Service/python/sns.py
# 
# IPWorks MQ 2024 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of IPWorks MQ in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/ipworksmq
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from ipworksmq import *

input = sys.hexversion < 0x03000000 and raw_input or input


def ensureArg(args, prompt, index):
    if len(args) <= index:
        while len(args) <= index:
            args.append(None)
        args[index] = input(prompt)
    elif args[index] is None:
        args[index] = input(prompt)


sns = AmazonSNS()
buffer = ""

def parseArgs(argv: list):
    maped = {}
    
    argc = len(argv)
    for i in range(argc):
        if argv[i][0] == "-":
            if i + 1 < argc and (argv[i + 1][0] != "-"):
                argv[i] = argv[i][1:]
                maped[argv[i].lower()] = argv[i + 1]
                i += 1
            else:
                argv[i] = argv[i][1:]
                maped[argv[i].lower()] = ""
        else:
            maped[i] = argv[i]

    return maped

def print_menu():
    print("Menu options:")
    print("? - Show menu")
    print("lst - List topics")
    print("lss - List subscriptions")
    print("lsst - List subscriptions by topic")
    print("mk - Create topic")
    print("del - Delete topic")
    print("sub - Subscribe to topic")
    print("unsub - Unsubscribe from topic")
    print("pub - Publish message")
    print("q - Quit")

def subscription_list(amazon_sns_subscription_list_event: AmazonSNSSubscriptionListEventParams):
    print("\ntopic ARN:", amazon_sns_subscription_list_event.topic_arn)
    print("subscription ARN:", amazon_sns_subscription_list_event.subscription_arn)
    print("endpoint:", amazon_sns_subscription_list_event.endpoint)
    print("owner:", amazon_sns_subscription_list_event.owner)
    
    if amazon_sns_subscription_list_event.protocol == 0:
        print("EndPoint protocol: email")
    elif amazon_sns_subscription_list_event.protocol == 1:
        print("EndPoint protocol: email-json")
    elif amazon_sns_subscription_list_event.protocol == 2:
        print("EndPoint protocol: http")
    elif amazon_sns_subscription_list_event.protocol == 3:
        print("EndPoint protocol: https")
    elif amazon_sns_subscription_list_event.protocol == 4:
        print("EndPoint protocol: SMS")
    elif amazon_sns_subscription_list_event.protocol == 5:
        print("EndPoint protocol: SQS")
    else:
        print("not known protocol")
    
    print("-----------------------------------------------------------------------------------------")

def topic_list(amazon_sns_topic_list_event: AmazonSNSTopicListEventParams):
    print(amazon_sns_topic_list_event.topic_arn)

sns.on_topic_list = topic_list
sns.on_subscription_list = subscription_list

try:
    if len(sys.argv) < 4:
        print("usage: sns.py -a Access_Key -s Secret_Key \n" \
            "Access_Key: the access key found on your AWS console\n" \
            "Secret_Key: the secret key generated after the access key\n")
        
        sys.exit(0)

    args = parseArgs(sys.argv)
    sns.access_key = args["a"]
    sns.secret_key = args["s"]

    print_menu()
    while True:
        buffer = input("Enter command: ").lower()

        try:
            if buffer == "?":
                print_menu()
            elif buffer == "lst":
                sns.list_topics()
                while sns.get_topic_marker() != "":
                    sns.list_topics()

            elif buffer == "lss":
                sns.list_subscriptions()
                while sns.get_subscription_marker() != "":
                    sns.list_subscriptions()

            elif buffer == "lsst":
                sns.list_topic_subscriptions(input("Enter topic ARN: "))

            elif buffer == "mk":
                print("Topic ARN for new topic: " + sns.create_topic(input("Enter topic name: ")))

            elif buffer == "del":
                sns.delete_topic(input("Enter topic ARN: "))
                print("success!")

            elif buffer == "sub":
                topic_arn = input("Enter a topic: ")
                endpoint = input("Enter an endpoint i.e email add, phone nr..: ")                
                endpoint_protocol = int(input("Choose a endpoint protocol: \n" +
                      "0 for email\n" +
                      "1 for email-json\n" +
                      "2 for http\n" +
                      "3 for https\n" +
                      "4 for sms\n" +
                      "5 for sqs\n"))
                
                print("Subscription ARN for new subscriber: " + sns.subscribe(topic_arn, endpoint, endpoint_protocol))

            elif buffer == "unsub":
                sns.unsubscribe(input("Subscription ARN: "))
                print("success!")

            elif buffer == "pub":
                topic_arn1 = input("Enter a topic: ")
                subject = input("Enter subject: ")
                message = input("Enter body: ")
                
                print("Message Id: " + sns.publish(topic_arn1, subject, message))

            elif buffer == "q":
                break

            else:
                print("not an option!")
                print_menu()

        except IPWorksMQError as e:
            print(e)

except Exception as e:
    print(e)

File: IPWorks MQ Samples/Amazon SNS Notification Service/c++/sns.cpp
/*
 * IPWorks MQ 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks MQ in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksmq
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <string>
#include <stdlib.h>
#include <unordered_map>
#include "../../include/amazonsns.h"
#define MAX_LEN 500

class MySNS : public AmazonSNS
{
public:
  virtual int FireTopicList(AmazonSNSTopicListEventParams* e)
  {
    printf("%s\n", e->TopicArn);
    return 0;
  }

  virtual int FireSubscriptionList(AmazonSNSSubscriptionListEventParams* e)
  {
    printf("\ntopic ARN: %s\n", e->TopicArn);
    printf("subscription ARN: %s\n", e->SubscriptionArn);
    printf("endpoint: %s\n", e->Endpoint);
    printf("owner: %s\n", e->Owner);

    switch (e->Protocol) {
    case 0:
      printf("EndPoint protocol: email\n");
      break;
    case 1:
      printf("EndPoint protocol: email-json\n");
      break;
    case 2:
      printf("EndPoint protocol: http\n");
      break;
    case 3:
      printf("EndPoint protocol: https\n");
      break;
    case 4:
      printf("EndPoint protocol: SMS\n");
      break;
    case 5:
      printf("EndPoint protocol: SQS\n");
      break;
    default:
      printf("not known protocol\n");
      break;
    }

    printf("-----------------------------------------------------------------------------------------\n");

    return 0;
  }

  virtual int FireError(AmazonSNSErrorEventParams* e)
  {
    printf("Error: %s\n\n", e->Description);
    return 0;
  }
};

void printMenu();
void toLowerCase(char* str);
std::unordered_map<std::string, std::string> parseArgs(int argc, char* argv[]);

int main(int argc, char* argv[])
{
  MySNS sns;
  char buffer[MAX_LEN];

  if (argc < 4)
  {
    printf("usage: sns.cpp -a Access_Key -s Secret_Key \n" \
      "Access_Key: the access key found on your AWS console\n" \
      "Secret_Key: the secret key generated after the access key\n");

    return 0;
  }

  std::unordered_map<std::string, std::string> map = parseArgs(argc, argv);

  sns.SetAccessKey(map.at("a").c_str());
  sns.SetSecretKey(map.at("s").c_str());

  printMenu();

  while (true)
  {
    int retCode = 0;
    printf("\nEnter command: ");
    scanf("%[^\n]", buffer);

    if (!strcmp(buffer, "lst"))
    {
      do
      {
        retCode = sns.ListTopics();
        if (retCode != 0)
        {
          printf("Error: [%i] %s\n\n", retCode, sns.GetLastError());
          break;
        }
      } while (strlen(sns.GetTopicMarker()) != 0);
    }
    else if (!strcmp(buffer, "lss"))
    {
      do
      {
        retCode = sns.ListSubscriptions();
        if (retCode != 0)
        {
          printf("Error: [%i] %s\n\n", retCode, sns.GetLastError());
          break;
        }
      } while (strlen(sns.GetSubscriptionMarker()) != 0);
    }
    else if (!strcmp(buffer, "lsst"))
    {
      printf("Enter topic ARN: ");
      getchar();
      scanf("%[^\n]", buffer);

      retCode = sns.ListTopicSubscriptions(buffer);
      if (retCode != 0)
      {
        printf("Error: [%i] %s\n\n", retCode, sns.GetLastError());
      }
    }
    else if (!strcmp(buffer, "mk"))
    {
      char* topicArn;
      printf("Enter topic name: ");
      getchar();
      scanf("%[^\n]", buffer);

      topicArn = sns.CreateTopic(buffer);
      if (strlen(topicArn) == 0)
      {
        getchar();
        printf("Something went wrong!");
        continue;
      }
      printf("Topic ARN for new topic: %s\n", topicArn);
    }
    else if (!strcmp(buffer, "del"))
    {
      printf("Enter topic ARN: ");
      getchar();
      scanf("%[^\n]", buffer);
      retCode = sns.DeleteTopic(buffer);
      if (retCode != 0)
      {
        printf("Error: [%i] %s\n\n", retCode, sns.GetLastError());
      }
    }
    else if (!strcmp(buffer, "sub"))
    {
      char topicArn[MAX_LEN];
      char endpoint[MAX_LEN];
      char* subArn;
      int endpointProtocol;

      printf("Enter a topic: ");
      getchar();
      scanf("%[^\n]", topicArn);

      printf("Enter an endpoint i.e email add, phone nr..: ");
      getchar();
      scanf("%[^\n]", endpoint);

      printf("Choose a endpoint protocol: \n" \
        "0 for email\n" \
        "1 for email-json\n" \
        "2 for http\n" \
        "3 for https\n" \
        "4 for sms\n" \
        "5 for sqs\n");

      scanf("%d", &endpointProtocol);

      subArn = sns.Subscribe(topicArn, endpoint, endpointProtocol);
      if (strlen(subArn) == 0)
      {
        printf("Something went wrong!");
        getchar();
        continue;
      }
      printf("Subscription ARN for new subscriber: %s\n", subArn);
    }
    else if (!strcmp(buffer, "unsub"))
    {
      printf("Subscription ARN: ");
      getchar();
      scanf("%[^\n]", buffer);
      retCode = sns.Unsubscribe(buffer);
      if (retCode != 0)
      {
        printf("Error: [%i] %s\n\n", retCode, sns.GetLastError());
      }
    }
    else if (!strcmp(buffer, "pub"))
    {
      char topicArn[MAX_LEN];
      char subject[MAX_LEN];
      char message[MAX_LEN];
      char* msgId;

      printf("Enter a topic: ");
      getchar();
      scanf("%[^\n]", topicArn);

      printf("Enter subject: ");
      getchar();
      scanf("%[^\n]", subject);

      printf("Enter body: ");
      getchar();
      scanf("%[^\n]", message);

      msgId = sns.Publish(topicArn, subject, message);
      if (strlen(msgId) == 0)
      {
        printf("Something went wrong!");
        getchar();
        continue;
      }
      printf("Message Id: %s", msgId);
    }
    else if (!strcmp(buffer, "q"))
    {
      return 0;
    }
    else if (!strcmp(buffer, "?"))
    {
      printMenu();
    }
    else
    {
      printf("not an option!");
      printMenu();
    }
    getchar();
  }
  return 0;
}

void printMenu()
{
  printf("\r\n?\t-\tHelp\n" \
    "lst\t-\tList Topics\n" \
    "lss\t-\tList Subscriptions\n" \
    "lsst\t-\tList Subscriptions by topic\n" \
    "mk\t-\tCreate Topic\n" \
    "del\t-\tDelete Topic\n" \
    "sub\t-\tSubscribe\n" \
    "unsub\t-\tUnSubscribe\n" \
    "pub\t-\tPublish\n" \
    "q\t-\tQuit\n");
}

std::unordered_map<std::string, std::string> parseArgs(int argc, char* argv[])
{
  std::unordered_map<std::string, std::string> map;
  for (int i = 0; i < argc; i++)
  {
    if (argv[i][0] == '-')
    {
      if (i + 1 < argc && (argv[i + 1][0] != '-'))
      {
        char* newStr = argv[i] + 1;
        toLowerCase(newStr);
        map[newStr] = argv[i + 1];
        i++;
      }
      else
      {
        char* newStr = argv[i] + 1;
        toLowerCase(newStr);
        map[newStr] = "";
      }
    }
    else
    {
      map[std::to_string(i)] = argv[i];
    }
  }
  return map;
}

void toLowerCase(char* str)
{
  for (size_t i = 0; i < std::strlen(str); ++i) {
    str[i] = std::tolower(static_cast<unsigned char>(str[i]));
  }
}

